lib/shared/bucketing-assembly-script/assembly/eventQueue/requestPayloadManager.ts

```diff
@@ -101,32 +101,59 @@ export class RequestPayloadManager {
             }
         }
 
-        // Generate defaulted agreagate user as our events APIs require a user / user_id
+        // Generate defaulted aggregate user as our events APIs require a user / user_id
         const dvcUser = new DVCPopulatedUser(new DVCUser(
             user_id, null, null, null, null, NaN, null, null, null, null
         ))
         this.addEventsToPendingPayloads(new UserEventsBatchRecord(dvcUser, aggEvents))
     }
 
+    /**
+     * Find existing FlushPayload that contains records for the user_id,
+     * or return the smallest FlushPayload or a new FlushPayload
+     */
+    private getFlushPayloadForUser_id(user_id: string): FlushPayload {
+        const keys = this.pendingPayloads.keys()
+        let smallestPayload: FlushPayload | null = null
+
+        for (let i=0; i < keys.length; i++) {
+            const key = keys[i]
+            const payload = this.pendingPayloads.get(key)
+            const recordForUser = payload.recordForUser(user_id)
+            const payloadEventCount = payload.eventCount()
+            if (recordForUser && payloadEventCount < this.chunkSize) {
+                this.pendingPayloads.delete(key)
+                return payload
+            }
+            if (payloadEventCount < this.chunkSize
+                && (!smallestPayload || smallestPayload.eventCount() < payloadEventCount)) {
+                smallestPayload = payload
+            }
+        }
+
+        if (smallestPayload) {
+            this.pendingPayloads.delete(smallestPayload.payloadId)
+            return smallestPayload
+        } else {
+            return new FlushPayload([])
+        }
+    }
+
     /**
      * Chunk up UserEventsBatchRecord's into payloads of size: this.chunkSize
      */
     private addEventsToPendingPayloads(record: UserEventsBatchRecord): void {
-        let flushPayload = new FlushPayload([])
+        let flushPayload = this.getFlushPayloadForUser_id(record.user.user_id)
 
-        while (record.events.length > this.chunkSize) {
-            const batchRecord = new UserEventsBatchRecord(
-                record.user,
-                record.events.splice(0, this.chunkSize)
-            )
-            flushPayload.records.push(batchRecord)
+        while (record.events.length > (this.chunkSize - flushPayload.eventCount())) {
+            flushPayload.addBatchRecordForUser(record, this.chunkSize)
             this.pendingPayloads.set(flushPayload.payloadId, flushPayload)
             flushPayload = new FlushPayload([])
         }
         if (record.events.length > 0) {
-            flushPayload.records.push(record)
+            flushPayload.addBatchRecordForUser(record, this.chunkSize)
         }
-        if (flushPayload.records.length > 0) {
+        if (flushPayload.eventCount() > 0) {
```

Comment:
this is kind of confusing, are you potentially constructing flushPayload objects that dont end up getting used and therefore need to check if they have any events before adding to the pendingPayloads map?